aten/src/ATen/native/Convolution.cpp

```diff
@@ -161,6 +161,8 @@ auto ConvParams::use_cpu_depthwise3x3_winograd(
              (bias.scalar_type() == at::kFloat))) &&
          !is_strided() &&
          !is_dilated() &&
+         // 3x3 depthwith convolutions implementation is inference only
+         !(input.requires_grad() && GradMode::is_enabled()) &&
```

Comment:
Yeah, if GradMode is not enabled, we should be ok (no backward path is expected), but we also should be checking if weight requires grad (e.g. in the very first layer input doesn't require grad, but we still can't use this kernel)

Reply:
Yes, I think `!GradMode::is_enabled() || (GradMode::is_enabled() && !(input.requires_grad() || weight.requires_grad() || bias.requires_grad()))` is the right condition here.

Reply:
Second `GradMode::is_enabled()` is unneeded (if we moved on to or branch, it's true)

Reply:
Thank you for the suggestions. Amended. I still prefer `!(a && b)` syntax to `!a || !b` and also it makes short-circuiting rule a bit more explicit, i.e.:
```
 !(GradMode::is_enabled() && 
  (input.requires_grad() ||
   weight.requires_grad() ||
   (bias.defined() && bias.requires_grad())))
```